ykrt/pass_finder/try_passes.py

```diff
@@ -185,18 +211,114 @@ def binary_split(logf, passes, is_prelink):
         print(72 * "=")
         print(list_of_passes_to_str(ok_passes))
 
-def main(logf, is_prelink):
+def evaluate_fitness(glogf, is_prelink, entity, passes, cwd):
+    try_passes = []
+    for i, bit in enumerate(entity):
+        if bit == 1:
+            try_passes.append(passes[i])
+    log(glogf, f"\ncurrently evaluating {try_passes}\n")
+    config = get_pipeline_config(is_prelink, try_passes)
+    ret, exec_time = test_pipeline(glogf, config, cwd) 
+    if ret:
+            exec_time = float(exec_time)  # Convert exec_time to a float
+            return  exec_time
+    else:
+        # Return False for programs that crash, and execution time as a large positive value
+        return float('inf')
+
+def crossover(parent1, parent2): 
+    crossover_point = random.randint(1, len(parent1) - 1)
+    child1 = parent1[:crossover_point] + parent2[crossover_point:]
+    child2 = parent2[:crossover_point] + parent1[crossover_point:]
+    return child1, child2
+
+def mutate(entity, mutation_rate):
+    mutated_entity = []
+    for bit in entity:
+        if random.random() < mutation_rate:
+            mutated_entity.append(1 - bit)  # Flip the bit
+        else:
+            mutated_entity.append(bit)
+    return mutated_entity
+
+def genetic_algorithm(glogf, is_prelink, population_size, mutation_rate, generations, target_fitness, passes, cwd):
+    population = []
+    fitness_scores = []
+
+    for _ in range(population_size):
+        entity = [random.randint(0,1) for _ in range(len(passes))]
+        population.append(entity)
+
+    for generation in range(generations):
+        log(glogf, "=========================================================")
+        log(glogf, f"\ngeneration: {generation}\n") 
+        fitness_scores.clear()
+        # Evaluate fitness for each entity in the population
+        fitness_scores = [evaluate_fitness(glogf, is_prelink, entity, passes, cwd) for entity in population]
+        log(glogf, f"\nfitness score: {fitness_scores}\n")
+        log(glogf, "=========================================================")
+
+        # less the execution time is better
+        wt = [(1/t) for t in fitness_scores] 
+        print(f"\033[38;5;128m {fitness_scores}\033[0m")
+        
+        # Check if we have reached the target fitness
+        if target_fitness in fitness_scores:
+            print(f"Target fitness reached in generation {generation + 1}!")
+            break
+        
+        # to randomly pick entities when the complete population fails
+        if fitness_scores.count(float('inf')) == len(fitness_scores):
+            wt = [1 for _ in fitness_scores]
+
+        # Select parents for reproduction (roulette wheel selection)
+        parents = []
+        for _ in range(population_size // 2):
+            parent1 = random.choices(population, weights=wt, k=1)[0]
+            parent2 = random.choices(population, weights=wt, k=1)[0]
```

Comment:
Is it a requirement for roulette wheel that the parents are different? I dunno!

Reply:
we'll need seperarte entity for crossover and mutation. That's why they are different.

Reply:
Since they're randomly chosen, can't `parent1` and `parent2` end up being the same value?

Reply:
They can but not always (if I dare to say), however I think it is important for us to have `parent1` and `parent2` because the `crossover` function finds a pivot and takes first section (till pivot point) from `parent1` and rest from `parent2`. 
Also, these parents are appended in a parents list, which is then used to get children by crossover-ing and mutating.
I am not sure how we are supposed to operate `crossover` function if both parent are always same.

Reply:
So if the parents must be different, we have to do something to make that happen. For example, select parent1 then in a loop select a parent2 until parent2 != parent1.

Reply:
Changed here: 0edbc0d87347d1d384b8dd4545f5e0d77115fbcf